Exiled.CustomRoles/API/Features/CustomRole.cs

```diff
@@ -0,0 +1,445 @@
+// -----------------------------------------------------------------------
+// <copyright file="CustomRole.cs" company="Exiled Team">
+// Copyright (c) Exiled Team. All rights reserved.
+// Licensed under the CC BY-SA 3.0 license.
+// </copyright>
+// -----------------------------------------------------------------------
+
+namespace Exiled.CustomRoles.API.Features
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    using Dissonance.Integrations.MirrorIgnorance;
+
+    using Exiled.API.Enums;
+    using Exiled.API.Extensions;
+    using Exiled.API.Features;
+    using Exiled.API.Features.Items;
+    using Exiled.API.Features.Spawn;
+    using Exiled.CustomItems;
+    using Exiled.CustomItems.API.Features;
+    using Exiled.Events.EventArgs;
+    using Exiled.Loader;
+
+    using MEC;
+
+    using NorthwoodLib.Pools;
+
+    using UnityEngine;
+
+    using YamlDotNet.Serialization;
+
+    using Ragdoll = Ragdoll;
+
+    /// <summary>
+    /// The custom role base class.
+    /// </summary>
+    public abstract class CustomRole
+    {
+        /// <summary>
+        /// Gets a list of all registered custom roles.
+        /// </summary>
+        public static HashSet<CustomRole> Registered { get; } = new HashSet<CustomRole>();
+
+        /// <summary>
+        /// Gets or sets the custom RoleID of the role.
+        /// </summary>
+        public abstract uint Id { get; set; }
```

Comment:
Please don't do this. This reminds me of hellish modded minecraft configs where every id was numeric, and you had to configure each mod to not have conflicting IDs by going through every single config file. Why not simply have string names, or identify by plugin name + numerical id?

Reply:
Exiled.CustomItems already does this and there's been no complaints?
Technically the ID numbers could just be removed completely, but they are there to allow, for example, a command that gives a role to a player, that can accept the id # instead of the role's name. This makes life simpler for the creator of said commnad, because they don't need to parse the name etc, they can just use an int.

Reply:
My main concern here was that two plugins would register under the same id, and not have an option to change them (as is common with commands). To prevent that, there should be a rule requiring the item/role to be put inside the config, as is used in CustomRoles.